File: Wire-iOS/Sources/UserInterface/Components/KeyboardAvoidingViewController.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "KeyboardAvoidingViewController.h"
#import "KeyboardAvoidingViewController+Internal.h"

@import PureLayout;

#import "UIView+Zeta.h"
#import "Constants.h"
#import "KeyboardFrameObserver+iOS.h"
#import "Wire-Swift.h"



@interface KeyboardAvoidingViewController ()

@property (nonatomic, readwrite) UIViewController *viewController;
@property (nonatomic) NSLayoutConstraint *topEdgeConstraint;

@end



@implementation KeyboardAvoidingViewController

+ (void)load
{
    @autoreleasepool {
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            [KeyboardFrameObserver sharedObserver];
        });
    }
}

- (instancetype)initWithViewController:(UIViewController *)viewController
{
    self = [super initWithNibName:nil bundle:nil];
    
    if (self) {
        self.viewController = viewController;

        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(keyboardFrameWillChange:)
                                                     name:UIKeyboardWillChangeFrameNotification
                                                   object:nil];
    }
    
    return self;
}

- (BOOL)shouldAutorotate
{
    return [self.viewController shouldAutorotate];
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations
{
    return [self.viewController supportedInterfaceOrientations];
}

- (UINavigationItem *)navigationItem
{
    return [self.viewController navigationItem];
}

- (BOOL)prefersStatusBarHidden
{
    return self.viewController.prefersStatusBarHidden;
}

- (UIStatusBarStyle)preferredStatusBarStyle
{
    return self.viewController.preferredStatusBarStyle;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.view.opaque = NO;
    [self addChildViewController:self.viewController];
    [self.view addSubview:self.viewController.view];
    self.viewController.view.translatesAutoresizingMaskIntoConstraints = NO;
    [self.viewController didMoveToParentViewController:self];
    
    [self createInitialConstraints];
}

- (void)createInitialConstraints
{
    [self.viewController.view autoPinEdgeToSuperviewEdge:ALEdgeLeft withInset:0];
    [self.viewController.view autoPinEdgeToSuperviewEdge:ALEdgeRight withInset:0];
    self.topEdgeConstraint = [self.viewController.view autoPinEdgeToSuperviewEdge:ALEdgeTop withInset:self.topInset];
    
    self.bottomEdgeConstraint = [self.viewController.bottomLayoutGuide.bottomAnchor constraintEqualToAnchor:self.bottomLayoutGuide.bottomAnchor constant:0];
    [self.bottomEdgeConstraint setActive:YES];
}

- (void)setTopInset:(CGFloat)topInset
{
    _topInset = topInset;
    self.topEdgeConstraint.constant = topInset;
    [self.view setNeedsLayout];
}

- (NSString *)title
{
    return self.viewController.title;
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/Components/KeyboardAvoidingViewController.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

class KeyboardAvoidingViewController: UIViewController {
    let viewController: UIViewController
    var disabledWhenInsidePopover = false

    private var animator: UIViewPropertyAnimator?
    private var bottomEdgeConstraint: NSLayoutConstraint?
    private var topEdgeConstraint: NSLayoutConstraint?

    required init(viewController: UIViewController) {
        self.viewController = viewController

        super.init(nibName: nil, bundle: nil)

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(keyboardFrameWillChange),
            name: UIWindow.keyboardWillChangeFrameNotification,
            object: nil
        )
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override var shouldAutorotate: Bool {
        viewController.shouldAutorotate
    }

    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        viewController.supportedInterfaceOrientations
    }

    override var navigationItem: UINavigationItem {
        viewController.navigationItem
    }

    override var childForStatusBarStyle: UIViewController? {
        viewController
    }

    override var childForStatusBarHidden: UIViewController? {
        viewController
    }

    override var title: String? {
        get {
            viewController.title
        }
        set {
            viewController.title = newValue
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        view.isOpaque = false
        addChild(viewController)
        view.addSubview(viewController.view)
        view.backgroundColor = viewController.view.backgroundColor
        viewController.view.translatesAutoresizingMaskIntoConstraints = false
        viewController.didMove(toParent: self)

        createInitialConstraints()
    }

    private func createInitialConstraints() {
        NSLayoutConstraint.activate([
            viewController.view.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            viewController.view.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            viewController.view.topAnchor.constraint(equalTo: view.topAnchor),
        ])
        topEdgeConstraint = viewController.view.topAnchor.constraint(equalTo: view.topAnchor)
        topEdgeConstraint?.isActive = true

        bottomEdgeConstraint = viewController.view.safeAreaLayoutGuide.bottomAnchor
            .constraint(
                equalTo: view.safeAreaLayoutGuide.bottomAnchor,
                constant: 0
            )

        bottomEdgeConstraint?.isActive = true
    }

    @objc
    private func keyboardFrameWillChange(_ notification: Notification?) {
        guard let bottomEdgeConstraint else { return }

        guard !disabledWhenInsidePopover || !isInsidePopover else {
            bottomEdgeConstraint.constant = 0
            view.layoutIfNeeded()
            return
        }

        guard let userInfo = notification?.userInfo,
              let duration = userInfo[UIResponder.keyboardAnimationDurationUserInfoKey] as? TimeInterval
        else { return }

        let keyboardFrameInView = UIView.keyboardFrame(in: view, forKeyboardNotification: notification)
        var bottomOffset: CGFloat

        // The keyboard frame includes the safe area so we need to substract it since the bottomEdgeConstraint is
        // attached to the safe area.
        bottomOffset = -keyboardFrameInView.intersection(view.safeAreaLayoutGuide.layoutFrame).height

        // When the keyboard is visible &
        // this controller's view is presented at a form sheet style on iPad, the view is has a top offset and the
        // bottomOffset should be reduced.
        if !keyboardFrameInView.origin.y.isInfinite,
           modalPresentationStyle == .formSheet,
           let frame = presentationController?.frameOfPresentedViewInContainerView {
            // swiftlint:disable:next todo_requires_jira_link
            // TODO: no need to add when no keyboard
            bottomOffset += frame.minY
        }

        guard bottomEdgeConstraint.constant != bottomOffset else { return }

        // When the keyboard is dismissed and then quickly revealed again, then
        // the dismiss animation will be cancelled.
        animator?.stopAnimation(true)
        view.layoutIfNeeded()

        animator = UIViewPropertyAnimator(duration: duration, timingParameters: UISpringTimingParameters())

        animator?.addAnimations {
            bottomEdgeConstraint.constant = bottomOffset
            self.view.layoutIfNeeded()
        }

        animator?.addCompletion { [weak self] _ in
            self?.animator = nil
        }

        animator?.startAnimation()
    }
}
